network-config-analyzer/IstioSidecarYamlParser.py

```diff
@@ -0,0 +1,195 @@
+        """
+        if is_default_sidecar:
+            if self.namespace.ordered_default_sidecars:  # this sidecar is not first one
+                self.warning(f'Namespace "{str(self.namespace)}" already has a Sidecar configuration '
+                             f'without any workloadSelector. '
+                             f'Connections in sidecar: "{sidecar_full_name}" will be ignored')
+            self.namespace.ordered_default_sidecars.append(sidecar_full_name)
+            return
+
+        for peer in selected_peers:
+            if peer.ordered_specific_sidecars:  # this sidecar is not first one
+                self.warning(f'Peer "{peer.full_name()}" already has a Sidecar configuration selecting it.'
+                             f'Sidecar: "{sidecar_full_name}" will not be considered as connections for this workload')
+            peer.ordered_specific_sidecars.append(sidecar_full_name)
+
+    def parse_policy(self):
+        """
+        Parses the input object to create a IstioSidecar object
+        :return: a IstioSidecar object with proper PeerSets
+        :rtype: IstioSidecar
+        """
+        policy_name = self.parse_generic_istio_policy_fields('Sidecar', 'networking.istio.io/v1beta1')
+        if policy_name is None:
+            return None  # not relevant to build this policy
+
+        res_policy = IstioSidecar(policy_name, self.namespace)
+        res_policy.policy_kind = NetworkPolicy.PolicyType.IstioSidecar
+
+        if 'spec' not in self.policy or self.policy['spec'] is None:
+            self.warning('spec is missing or null in Sidecar ' + res_policy.full_name())
+            return res_policy
+
+        sidecar_spec = self.policy['spec']
+        # currently, supported fields in spec are workloadSelector and egress
+        allowed_spec_keys = {'workloadSelector': [0, dict], 'ingress': [0, list], 'egress': [0, list],
+                             'outboundTrafficPolicy': [2, str]}
+        self.check_fields_validity(sidecar_spec, 'Sidecar spec', allowed_spec_keys)
+        res_policy.affects_egress = sidecar_spec.get('egress') is not None
+
+        workload_selector = sidecar_spec.get('workloadSelector')
+        if workload_selector is None:
+            res_policy.default_sidecar = True
+        res_policy.selected_peers = self.update_policy_peers(workload_selector, 'labels')
+        self._append_sidecar_into_relevant_list(str(res_policy), res_policy.selected_peers, res_policy.default_sidecar)
+
+        if sidecar_spec.get('ingress') is not None:
+            self.warning('Sidecar ingress is not supported yet.'
+                         f'Although "{res_policy.full_name()}" contains ingress entry, '
+                         f'it is not considered in policy connections', sidecar_spec)
+
+        egress = sidecar_spec.get('egress', [])
+        if egress is None:
+            self.syntax_error('An empty configuration is provided', res_policy.name)  # behavior of live cluster
```

Comment:
removed the syntax error, but still saving the specific sidecar as priority regardless it includes "egress" or not because:
While the spec states that, it seems to me that the live-cluster behaves differently.
This is what I've tested on live-cluster:
I've applied first a default sidecar on namespace : default
[def_sidecar_default_namespace1.txt](https://github.com/IBM/network-config-analyzer/files/9439292/def_sidecar_default_namespace1.txt)
and connections from this were really applied on all apps ` including ratings app`
Then I applied a new sidecar selecting the `ratings app` without egress specified, only ingress.
[sidecar_only_ingress.txt](https://github.com/IBM/network-config-analyzer/files/9439307/sidecar_only_ingress.txt)
but this specified sidecar "cancelled" the "egress" connections of the default sidecar. (no other sidecars or policies were applied).

![image](https://user-images.githubusercontent.com/82180114/187064392-3541950f-c9c0-4ca7-8974-371a62616d5c.png)

I don't know whether my istio version is not updated. 
If we want to behave as written in the spec even though, I will call the `self._check_and_save_sidecar_if_top_priority(res_policy)` only if the sidecar contains egress rules. Should I do that?

Reply:
yes, the SideCar with no egress surely did something...  although I'm not sure I understand the output of `istioctl proxy-config clusters`. Did you also check the effect on connectivity? You may need to open a shell on the `ratings` container in the `ratings` Pod, and try for example to access the `reviews` service.

Reply:
It seems the output of `istioctl proxy-config clusters` should be the list of services handled by the proxy for a pod. So it makes sense from this output that the rules from the 'egress sidecar' are not affecting the proxy of this pod after applying the 'ingress sidecar' (which does seem contradictory with spec definition). I do think it is worth checking actual connectivity as you suggested, and whether the sidecar ingress rule has some possible effect (maybe try as well an alternative sidecar with no ingress/egress rules instead of the one with ingress rule ).

Reply:
- trying with another sidecar with a workloadSelector but without ingress and egress, got the same - the new sidecar overrides the whole default sidecar (the default's sidecar's egress is cancelled)

- testing connectivity on the shell did the same for me, the ratings pod could connect to services other than the ones defined in the default's sidecar egress 

It seems that a defined sidecar would override the default/global sidecar in any case (ignoring the existence of the 'egress' field ) - so leaving the code as is for now.

I have also published this contradicting "issue" on istio discussing forums. will update if I get a different answer/explanation

Reply:
Hi @zivnevo , @adisos ,
I got an answer in the following forum : [istio-discussion](https://discuss.istio.io/t/istio-sidecar-doc-description-vs-live-cluster-behavior/13849)

So from this answer I understand that the current code is fine (a specific sidecar always overrides a default one, regardless the existence of the 'egress' field)

Reply:
ok, please update the comment in the code with a link to this clarification.